File: core/src/main/java/com/nfthub/core/entity/Magazine.java
package com.nfthub.core.entity;

import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table
@Getter
@NoArgsConstructor
public class Magazine extends BaseTimeEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(length = 100)
    private String title;

    @Column(length = 200)
    private String description;

    @Column
    private String url;

    @Column
    private String author;

    @ManyToOne(fetch = FetchType.LAZY)
    private Category category;

    @OneToMany(mappedBy = "magazine", cascade = CascadeType.ALL)
    private List<MagazineTag> magazineTags = new ArrayList<>();

    @OneToMany(mappedBy = "magazine", cascade = CascadeType.ALL)
    private List<MagazineImage> magazineImages = new ArrayList<>();

    public Magazine(String title, String description, String url, String author) {
        this.title = title;
        this.description = description;
        this.url = url;
        this.author = author;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public void setCategory(Category category) {
        this.category = category;
    }
}

File: src/main/kotlin/com/nfthub/api/entity/Magazine.kt
package com.nfthub.api.entity

import com.nfthub.api.EMPTY_STRING
import javax.persistence.*

@Entity
@Table
class Magazine(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(nullable = false)
    var id: Long = 0,

    @Column(length = 100)
    var title: String = EMPTY_STRING,

    @Column(length = 200)
    var description: String = EMPTY_STRING,

    var url: String = EMPTY_STRING,

    var author: String = EMPTY_STRING,

    @ManyToOne(fetch = FetchType.LAZY, optional = true)
    var category: Category? = null,

    @OneToMany(mappedBy = "magazine", orphanRemoval = true, cascade = [CascadeType.ALL])
    var magazineTags: List<MagazineTag> = emptyList(),

    @OneToMany(mappedBy = "magazine", orphanRemoval = true, cascade = [CascadeType.ALL])
    var images: MutableList<MagazineImage> = mutableListOf(),

    ) : BaseTimeEntity()


@Entity
@Table
class MagazineTag(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(nullable = false)
    var id: Long = 0,
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    var magazine: Magazine = Magazine(),
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    var tag: Tag = Tag()
)

@Entity
@Table
class MagazineImage(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(nullable = false)
    var id: Long = 0,
    @Column(columnDefinition = "text", length = 500)
    var url: String = EMPTY_STRING,
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    var magazine: Magazine = Magazine(),
    var isMain: Boolean = false,
)